haystack/nodes/other/shaper.py

```diff
@@ -0,0 +1,364 @@
+            [d.content for d in docs]
+        )
+        tmp_registry: Dict[str, Callable] = {
+            "len": len,
+            "expand": (lambda expand_target, size: [expand_target] * size),
+            "expand:size": lambda documents: len(documents),  # pylint: disable=unnecessary-lambda
+            "concat": concat,
+            "concat_docs": concat_docs,
+        }
+        # see https://adamj.eu/tech/2022/01/05/how-to-make-immutable-dict-in-python/
+        self.function_registry = MappingProxyType(tmp_registry)
+
+    def invoke(self, parent, node: Dict[str, Any], invocation_context: Dict[str, Any]):
+        """
+        Invoke the function specified in the node and store the result in the invocation context.
+
+        If the node is a function invocation, invoke it with the provided arguments
+        For example, in the YAML snippet below, the function len is invoked with the argument documents,
+        the result is stored in the variable size, and then used in the next function invocation expand
+        ---
+            query:
+              func: expand
+              output: questions
+              params:
+                expand_target: query
+                size:
+                  func: len
+                  params:
+                    - documents
+        ---
+         Otherwise, if the node is an output variable, store the result in the invocation context
+         Assign the value of the input variable from the invocation context to the output variable, store the
+            output variable in the invocation context, and pass it to the next pipeline node
+            ____
+            - name: shaper
+             params:
+               inputs:
+                   query:
+                     output: questions
+        ---
+
+        """
+        if "func" in node:
+            self.invoke_function(parent, node, invocation_context)
+
+        elif "output" in node and parent in invocation_context:
+            invocation_context[node["output"]] = copy.deepcopy(invocation_context[parent])
+        else:
+            params_is_parent = "params" == parent
+            root_node = parent is None and all(key for key in node.keys() if key in invocation_context)
+            # if not a function, valid output, or params definition then this must be a root node in the inputs
+            # otherwise, something is wrong, raise an exception
+            if not params_is_parent and not root_node:
+                raise ValueError(
+                    f"Invalid input configuration. The input configuration should be a function invocation, or "
+                    f"an assignment. The input configuration is {node}"
+                )
+
+    def invoke_function(self, parent: str, node: Dict[str, Any], invocation_context: Dict[str, Any]):
+        """
+        Invoke a function on the given invocation context and stores the result back in the invocation context.
```

Comment:
I don't understand this. Does that mean that the function only operates on the `invocation_context` that is stored in meta and mutates that and then stores it back there? It does not change the actual pipeline inputs/outputs i.e. documents?

Why is it called invocation context?

Reply:
It operates on pipeline run parameters + intermediate results stored under meta["invocation_context"]. When I say invocation context, I mean the pipeline run parameters + meta["invocation_context"]